services/dbaas/dbaas_client.go

```diff
@@ -19,102 +19,245 @@ package dbaas
+		grpc.WithInsecure(),
+		grpc.WithConnectParams(grpc.ConnectParams{Backoff: backoffConfig, MinConnectTimeout: 10 * time.Second}),
+		grpc.WithUserAgent("pmm-managed/" + version.Version),
+	}
+
+	conn, err := grpc.DialContext(ctx, c.dbaasControllerAPIAddress, opts...)
+	if err != nil {
+		return errors.Errorf("failed to connect to dbaas-controller API: %v", err)
 	}
+	c.conn = conn
+
+	c.kubernetesClient = controllerv1beta1.NewKubernetesClusterAPIClient(conn)
+	c.xtradbClusterClient = controllerv1beta1.NewXtraDBClusterAPIClient(conn)
+	c.psmdbClusterClient = controllerv1beta1.NewPSMDBClusterAPIClient(conn)
+	c.logsClient = controllerv1beta1.NewLogsAPIClient(conn)
+
+	c.l.Info("Connected to dbaas-controller API.")
+	return nil
+}
+
+// Disconnect disconnects the client from dbaas-controller API.
+func (c *Client) Disconnect() error {
+	c.connM.Lock()
+	defer c.connM.Unlock()
+	c.l.Info("Disconnecting from dbaas-controller API.")
+	c.wg.Wait()
+	if c.conn == nil {
+		c.l.Warnf("Trying to disconnect from dbaas-controller API but the connection is not up.")
+		return nil
+	}
+
+	if err := c.conn.Close(); err != nil {
+		return errors.Errorf("failed to close conn to dbaas-controller API: %v", err)
+	}
+	c.conn = nil
+	c.l.Info("Disconected from dbaas-controller API.")
+	return nil
 }
 
 // CheckKubernetesClusterConnection checks connection with kubernetes cluster.
 func (c *Client) CheckKubernetesClusterConnection(ctx context.Context, kubeConfig string) (*controllerv1beta1.CheckKubernetesClusterConnectionResponse, error) {
-	return c.kubernetesClient.CheckKubernetesClusterConnection(ctx, &controllerv1beta1.CheckKubernetesClusterConnectionRequest{
-		KubeAuth: &controllerv1beta1.KubeAuth{Kubeconfig: kubeConfig},
-	})
+	c.connM.RLock()
+	defer c.connM.RUnlock()
+	c.wg.Add(1)
+
+	in := &controllerv1beta1.CheckKubernetesClusterConnectionRequest{
+		KubeAuth: &controllerv1beta1.KubeAuth{
+			Kubeconfig: kubeConfig,
+		},
+	}
+	out, err := c.kubernetesClient.CheckKubernetesClusterConnection(ctx, in)
+	c.wg.Done()
+	return out, err
 }
 
-// ListXtraDBClusters returns a list of XtraDB clusters.
 func (c *Client) ListXtraDBClusters(ctx context.Context, in *controllerv1beta1.ListXtraDBClustersRequest, opts ...grpc.CallOption) (*controllerv1beta1.ListXtraDBClustersResponse, error) {
-	return c.xtradbClusterClient.ListXtraDBClusters(ctx, in, opts...)
+	c.connM.RLock()
+	defer c.connM.RUnlock()
+	c.wg.Add(1)
+	resp, err := c.xtradbClusterClient.ListXtraDBClusters(ctx, in, opts...)
+	c.wg.Done()
+	return resp, err
 }
 
 // CreateXtraDBCluster creates a new XtraDB cluster.
 func (c *Client) CreateXtraDBCluster(ctx context.Context, in *controllerv1beta1.CreateXtraDBClusterRequest, opts ...grpc.CallOption) (*controllerv1beta1.CreateXtraDBClusterResponse, error) {
-	return c.xtradbClusterClient.CreateXtraDBCluster(ctx, in, opts...)
+	c.connM.RLock()
+	defer c.connM.RUnlock()
+	c.wg.Add(1)
+	resp, err := c.xtradbClusterClient.CreateXtraDBCluster(ctx, in, opts...)
```

Comment:
why are we using waitGroups here? are these function calls async? if not I guess we can remove it

Reply:
WaitGroup was removed.